### PR TITLE
ENH: new .agg for list-likes

### PR DESCRIPTION
**Pull request by rhshadrach**
Part of #43678

- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

Replaces the list-like implementation for `.agg` used by Series, DataFrame, SeriesGroupBy, DataFrameGroupBy, Window, and Resample. This will transpose agg results, and swap multiindex levels as well as the order from within the levels. Adds the new implementation behind the option "new_udf_methods".

Docs: [future_udf_behavior.zip](https://github.com/pandas-dev/pandas/files/7368201/future_udf_behavior.zip)

### CODE REVIEW

**Comment by attack68** [comment]
just a comment - I have always tried to avoid names like "new_routine" because at some point, it won't be, and you wished you never called it that in the first place! :)

**Comment by rhshadrach** [comment]
@attack68 - The plan is to put all usage of these methods behind the option "new_udf_methods"; when the old are deprecated and then removed, they will be renamed by dropping all the "new_" prefixes. If you have any suggestions for different names, please to make them!

**Comment by attack68** [comment]
no, no suggestions. But, and I kid you not, in my line of work I have seen `swap_pricer()`, `new_swap_pricer()`, `new_swap_pricer2()` and `new_new_swap_pricer()` - imagine explaining that to new recruits!

**Comment by mroeschke** [comment]
Agreed, with @attack68. Helpful to have a slightly clearer naming even during the transition period. Maybe `shared_list_like` if the plan is for `method` to be agg, transform, etc?

**Comment by rhshadrach** [comment]
@mroeschke and @attack68 - is there opposition to the "new_udf_methods" option as well, or is it just on the method names themselves? I am not a particular fan of "new" myself, but I could not think of anything concise and better. Regardless of the name that is picked, I would like the relationship between the option name and the methods that are behind that option name to be clear.

**Comment by mroeschke** [comment]
> I would like the relationship between the option name and the methods that are behind that option name to be clear

+1 I would hope these are consistent.

Maybe `future_udf_behavior` if the idea is that this will be the default some day?

**Comment by rhshadrach** [comment]
Thanks @mroeschke - I like "future".

File: pandas/core/apply.py [hunk 1]
```diff
@@ -408,6 +414,77 @@ def agg_list_like(self) -> DataFrame | Series:
             )
             return concatenated.reindex(full_ordered_index, copy=False)
 
+    def future_list_single_arg(
```
Comment (jbrockmendel) [inline pandas/core/apply.py:417]: same request from the other PR re naming; "future" won't be very helpful for a reader a year from now

Comment (rhshadrach) [reply pandas/core/apply.py:417]: The intention is to have "future_" methods alongside the current methods; all with the same prefix so they are easy to identify. Any such method is behind the option "future_udf_behavior" meaning they will only be called when set to True. Assuming we do end up going forward with this new (experimental) behavior, once it is in a good place we deprecate the option and then remove the option.

A year from now, we will still have the option "future_udf_behavior", and in my opinion, the "future_" prefix is meaningful and helpful. When the option is removed, the "old" methods are removed and the "future_" methods are renamed by removing the prefix (none are public).

Comment (jbrockmendel) [reply pandas/core/apply.py:417]: you've clearly given this more thought than i have so im going to stop complaining about this, will instead grumble to myself

Comment (rhshadrach) [reply pandas/core/apply.py:417]: The name is still up for improvements and suggestions are most welcome, but wanted to explain why I felt "future" was appropriate.

**Pull request closed without merging**